api/propagation/propagation.go

```diff
@@ -0,0 +1,143 @@
+type HTTPExtractor interface {
+	// Extract method retrieves encoded information using supplier
+	// from the associated carrier, decodes it and creates a new
+	// context containing the decoded information.
+	//
+	// Information can be a correlation context or a remote span
+	// context. In case of span context, the propagator should
+	// store it in the context using
+	// trace.ContextWithRemoteSpanContext. In case of correlation
+	// context, the propagator should use correlation.WithMap to
+	// store it in the context.
+	Extract(context.Context, HTTPSupplier) context.Context
+}
+
+// HTTPInjector injects information into a HTTPSupplier.
+type HTTPInjector interface {
+	// Inject method retrieves information from the context,
+	// encodes it into propagator specific format and then injects
+	// the encoded information using supplier into an associated
+	// carrier.
+	Inject(context.Context, HTTPSupplier)
+}
+
+// Config contains the current set of extractors and injectors.
+type Config struct {
+	httpEx []HTTPExtractor
+	httpIn []HTTPInjector
+}
+
+// Propagators is the interface to a set of injectors and extractors
+// for all supported carrier formats. It can be used to chain multiple
+// propagators into a single entity.
+type Propagators interface {
+	// HTTPExtractors returns the configured extractors.
+	HTTPExtractors() []HTTPExtractor
+
+	// HTTPInjectors returns the configured injectors.
+	HTTPInjectors() []HTTPInjector
+}
+
+// HTTPPropagator is the interface to inject to and extract from
+// HTTPSupplier.
+type HTTPPropagator interface {
+	HTTPInjector
+	HTTPExtractor
+
+	// GetAllKeys returns the HTTP header names used.
+	GetAllKeys() []string
+}
+
+// Option support passing configuration parameters to New().
+type Option func(*Config)
+
+// propagators is the default Propagators implementation.
+type propagators struct {
+	config Config
+}
+
+// New returns a standard Propagators implementation.
+func New(options ...Option) Propagators {
+	config := Config{}
+	for _, opt := range options {
+		opt(&config)
+	}
+	return &propagators{
+		config: config,
+	}
+}
+
+// WithInjectors appends to the optional injector set.
+func WithInjectors(inj ...HTTPInjector) Option {
+	return func(config *Config) {
+		config.httpIn = append(config.httpIn, inj...)
+	}
+}
+
+// WithExtractors appends to the optional extractor set.
+func WithExtractors(ext ...HTTPExtractor) Option {
+	return func(config *Config) {
+		config.httpEx = append(config.httpEx, ext...)
+	}
+}
+
+// HTTPExtractors implements Propagators.
+func (p *propagators) HTTPExtractors() []HTTPExtractor {
+	return p.config.httpEx
+}
+
+// HTTPExtractors implements Propagators.
+func (p *propagators) HTTPInjectors() []HTTPInjector {
+	return p.config.httpIn
+}
+
+// ExtractHTTP applies props.HTTPExtractors() to the passed context
+// and the supplier and returns the combined result context.
+func ExtractHTTP(ctx context.Context, props Propagators, supplier HTTPSupplier) context.Context {
```

Comment:
There is a WIP on this topic that suggests the `ExtractHTTP` and `InjectHTTP` static methods here should be exposed composite extractor or injector, i.e., suggests that instead `Propagators` should implement HTTPExtractor and HTTPInjector interfaces itself. 

https://github.com/open-telemetry/opentelemetry-specification/pull/440

Reply:
Please replace these static methods with `Propagator` methods.

Reply:
As discussed in the SIG, this will take a bit more work and it's OK to merge what we have here and do the rest in a future PR.